runtime/strategies/search-tokens-to-particles.js

```diff
@@ -13,50 +13,77 @@ import {Walker} from '../recipe/walker.js';
 export class SearchTokensToParticles extends Strategy {
   constructor(arc) {
     super();
-    // TODO: Recipes. Handles?
+
     this._particleByToken = {};
+    this._particleByPhrase = {};
     this._recipeByToken = {};
+    this._recipeByPhrase = {};
     for (let particle of arc.context.particles) {
-      let name = particle.name.toLowerCase();
-      this._addParticle(name, particle);
-
-      let verb = particle.primaryVerb;
-      if (verb != name) {
-        this._addParticle(verb, particle);
-      }
+      this._addThing(particle.name, particle, this._particleByToken, this._particleByPhrase);
+      this._addThing(particle.primaryVerb, particle, this._particleByToken, this._particleByPhrase);
     }
     for (let recipe of arc.context.recipes) {
-      if (recipe.name) {
-        this._addRecipe(recipe.name.toLowerCase(), recipe);
-      }
-      recipe.verbs.forEach(verb => {
-        this._addRecipe(verb, recipe);
-      });
+      this._addThing(recipe.name, recipe, this._recipeByToken, this._recipeByPhrase);
+      recipe.verbs.forEach(verb => this._addThing(verb, recipe, this._recipeByToken, this._recipeByPhrase));
     }
 
     let findParticles = token => this._particleByToken[token] || [];
     let findRecipes = token => this._recipeByToken[token] || [];
+    let particleByPhrase = this._particleByPhrase;
+    let recipeByPhrase = this._recipeByPhrase;
+
     class SearchWalker extends Walker {
       onRecipe(recipe) {
         if (!recipe.search || !recipe.search.unresolvedTokens.length) {
           return;
         }
 
         let byToken = {};
+        let resolvedTokens = new Set();
+
+        Object.keys(particleByPhrase).forEach(phrase => {
```

Comment:
There is a fair bit of duplication. I think we could unify some concepts here (recipe vs particle, token vs phrase).

We could have a single map (phrase => [thing]), where thing is {spec} or {innerRecipe} (why not simply {recipe}? ). phrase could be a post processed phrase with spaces (as you do below with regex) or a single token.

If there was a particle.verbs getter (why it doesn't exist? I'm confused, there is particle._verbs which is an array), we could write roughly below instead of 17-33:

let thingByPhrase = {};
for (let [thing, packaged] of [
    ...arc.context.particles.map(p => [p, {spec: p}]),
    ...arc.context.recipes.map(r => [r, {recipe: r}])]) {
  this._addThing(thing.name, packaged, thingByPhrase);
  thing.verbs.forEach(verb => this._addThing(verb, packaged, thingByPhrase));
}

Then we could replace lines 44-84 with something like:

for (let [phrase, thing] of Object.entries(thingByPhrase)) {
  let tokens = phrase.split(' '); // <- this maybe just an array of one if phrase is already a token.
  if (tokens.every(token => recipe.search.unresolvedTokens.find(unresolved => unresolved == token)) &&
      recipe.search.phrase.includes(phrase)) {
    byToken[phrase] = byToken[phrase] || [];
    byToken[phrase].push(thing);
    tokens.forEach(token => resolvedTokens.add(token));
  }
}

It would do more work for *ByToken lists, as it searches through entire population of particles/recipes instead of only looking up unresolved tokens, but we already do it for phrases, so complexity is the same. Seems to me that more concise code is worth it.

WDYT?

Reply:
Thanks for the suggestions, the changes I made are in between the original code and what you suggested, but it's already much better. I think, I still want to keep phrases and tokens separate, because phrases are higher priority, for example:
```
particle FindRestaurants &find in '...;
   ... 
particle FindProducts &find in '...'
   ...
recipe
  search `find restaurants`
```
I want to match FindRestaurants.